adala/utils/llm.py

```diff
@@ -0,0 +1,235 @@
+
+
+def get_messages(user_prompt: str, system_prompt: Optional[str] = None, instruction_first: bool = True):
+    messages = [{'role': 'user', 'content': user_prompt}]
+    if system_prompt:
+        if instruction_first:
+            messages.insert(0, {'role': 'system', 'content': system_prompt})
+        else:
+            messages[0]['content'] += system_prompt
+    return messages
+
+
+async def async_get_llm_response(
+    user_prompt: Optional[str] = None,
+    system_prompt: Optional[str] = None,
+    messages: Optional[List[Dict[str, str]]] = None,
+    model: Optional[str] = 'gpt-4o-mini',
+    instruction_first: Optional[bool] = True,
+    response_model: Optional[Type[BaseModel]] = None,
+    api_key: Optional[str] = None,
+    max_tokens: Optional[int] = 1000,
+    temperature: Optional[float] = 0.0,
+    timeout: Optional[Union[float, int]] = None,
+) -> Dict[str, Any]:
+    """
+    Async version of create_completion function with error handling and session timeout.
+
+    Args:
+        model: model name. Refer to litellm supported models for how to pass
+               this: https://litellm.vercel.app/docs/providers
+        user_prompt: User prompt.
+        system_prompt: System prompt.
+        messages: List of messages to be sent to the model. If provided, `user_prompt`, `system_prompt` and `instruction_first` will be ignored.
+        api_key: API key, optional. If provided, will be used to authenticate
+                 with the provider of your specified model.
+        instruction_first: Whether to put instructions first.
+        response_model: Pydantic model to constrain the LLM generated response. If not provided, the raw completion text will be returned.  # noqa
+        max_tokens: Maximum tokens to generate.
+        temperature: Temperature for sampling.
+        timeout: Timeout in seconds.
+
+    Returns:
+        Dict[str, Any]: OpenAI response or error message.
+    """
+
+    if not user_prompt and not messages:
+        raise ValueError("You must provide either `user_prompt` or `messages`.")
+
+    if not messages:
+        # get messages from user_prompt and system_prompt
+        messages = get_messages(user_prompt, system_prompt, instruction_first)
+    response = LLMResponse()
+
+    if response_model is None:
+        # unconstrained generation - return raw completion text and store it in `data` field: {"text": completion_text}
+        try:
+            completion = await litellm.acompletion(
+                model=model,
+                api_key=api_key,
+                messages=messages,
+                max_tokens=max_tokens,
+                temperature=temperature,
+                timeout=timeout,
+            )
+            completion_text = completion.choices[0].message.content
+            response.data['_completion_text'] = completion_text
+        except Exception as e:
+            response.adala_error = True
```

Comment:
I figure we can use the LLMResponse object for all of our internal logic, and hold off on doing the model_dump until we have to serialize it for something external. Maybe I'm missing something here though.

Reply:
Done with returning models:
```
class LLMResponse(BaseModel):
    """
    Base class for LLM response.
    """


class ConstrainedLLMResponse(LLMResponse):
    """
    LLM response from constrained generation.
    `data` object contains fields required by the response model.
    """
    data: Dict = Field(default_factory=dict)


class UnconstrainedLLMResponse(LLMResponse):
    """
    LLM response from unconstrained generation.
    `text` field contains raw completion text.
    """
    text: str = Field(default=None)


class ErrorLLMResponse(LLMResponse):
    """
    LLM response in case of error.
    """
    adala_error: bool = Field(default=True, serialization_alias='_adala_error')
    adala_message: str = Field(default=None, serialization_alias='_adala_message')
    adala_details: str = Field(default=None, serialization_alias='_adala_details')
```

however, logic with putting everything in dataframe is still there by design.